File: backend/app/clustering.py
import logging
from dataclasses import dataclass
from typing import List, Dict, Tuple

import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

from app.config import settings

logger = logging.getLogger(__name__)

SPANISH_STOP_WORDS = [
    "de", "la", "que", "el", "en", "y", "a", "los", "del", "se", "las",
    "por", "un", "para", "con", "una", "su", "al", "lo", "como", "más",
    "pero", "sus", "le", "ya", "o", "este", "sí", "porque", "esta", "entre",
    "cuando", "muy", "sin", "sobre", "también", "me", "hasta", "hay", "donde",
    "quien", "desde", "todo", "nos", "durante", "todos", "uno", "les", "ni",
    "contra", "otros", "ese", "eso", "ante", "ellos", "e", "esto", "mí",
    "antes", "algunos", "qué", "unos", "yo", "otro", "otras", "otra", "él",
    "tanto", "esa", "estos", "mucho", "quienes", "nada", "muchos", "cual",
    "poco", "ella", "estar", "estas", "alguno", "alguna", "aunque", "siempre",
    "fue", "ser", "es", "son", "han", "ha", "tiene", "tienen", "había",
    "argentina", "argentino", "argentinos", "argentina", "años", "año",
]


@dataclass
class ArticleInput:
    id: int
    title: str
    summary: str
    source_slug: str


@dataclass
class Cluster:
    article_ids: List[int]
    source_slugs: List[str]
    representative_title: str
    similarity_scores: Dict[int, float]


def cluster_articles(articles: List[ArticleInput], threshold: float = None) -> List[Cluster]:
    if threshold is None:
        threshold = settings.CLUSTERING_THRESHOLD

    if len(articles) < 2:
        return []

    texts = [f"{a.title} {a.summary}" for a in articles]

    vectorizer = TfidfVectorizer(
        stop_words=SPANISH_STOP_WORDS,
        ngram_range=(1, 2),
        max_features=5000,
        sublinear_tf=True,
    )
    try:
        tfidf_matrix = vectorizer.fit_transform(texts)
    except Exception as e:
        logger.error(f"TF-IDF vectorization failed: {e}")
        return []

    sim_matrix = cosine_similarity(tfidf_matrix)

    n = len(articles)
    visited = [False] * n
    clusters: List[Cluster] = []

    for i in range(n):
        if visited[i]:
            continue

        cluster_indices = [i]
        visited[i] = True

        for j in range(i + 1, n):
            if not visited[j] and sim_matrix[i][j] >= threshold:
                # Only cluster if from different sources
                sources_in_cluster = {articles[k].source_slug for k in cluster_indices}
                if articles[j].source_slug not in sources_in_cluster:
                    cluster_indices.append(j)
                    visited[j] = True

        # Only create cluster if 2+ different sources
        source_slugs = [articles[k].source_slug for k in cluster_indices]
        if len(set(source_slugs)) >= 2:
            scores = {articles[k].id: float(sim_matrix[i][k]) for k in cluster_indices}
            # Pick title from the article with most words (usually more complete)
            rep_idx = max(cluster_indices, key=lambda k: len(articles[k].title.split()))
            clusters.append(Cluster(
                article_ids=[articles[k].id for k in cluster_indices],
                source_slugs=source_slugs,
                representative_title=articles[rep_idx].title,
                similarity_scores=scores,
            ))

    logger.info(f"Clustered {len(articles)} articles into {len(clusters)} clusters")
    return clusters

File: frontend/src/lib/clustering.ts
/* CADA LADO — TF-IDF + cosine similarity clustering (TS port of backend/app/clustering.py) */

const SPANISH_STOP_WORDS = new Set([
  'de', 'la', 'que', 'el', 'en', 'y', 'a', 'los', 'del', 'se', 'las',
  'por', 'un', 'para', 'con', 'una', 'su', 'al', 'lo', 'como', 'más',
  'pero', 'sus', 'le', 'ya', 'o', 'este', 'sí', 'porque', 'esta', 'entre',
  'cuando', 'muy', 'sin', 'sobre', 'también', 'me', 'hasta', 'hay', 'donde',
  'quien', 'desde', 'todo', 'nos', 'durante', 'todos', 'uno', 'les', 'ni',
  'contra', 'otros', 'ese', 'eso', 'ante', 'ellos', 'e', 'esto', 'mí',
  'antes', 'algunos', 'qué', 'unos', 'yo', 'otro', 'otras', 'otra', 'él',
  'tanto', 'esa', 'estos', 'mucho', 'quienes', 'nada', 'muchos', 'cual',
  'poco', 'ella', 'estar', 'estas', 'alguno', 'alguna', 'aunque', 'siempre',
  'fue', 'ser', 'es', 'son', 'han', 'ha', 'tiene', 'tienen', 'había',
  'argentina', 'argentino', 'argentinos', 'años', 'año',
])

export interface ArticleInput {
  id:         number
  title:      string
  summary:    string
  sourceSlug: string
}

export interface ClusterResult {
  articleIds:        number[]
  sourceSlugs:       string[]
  representativeTitle: string
  similarityScores:  Record<number, number>
}

/* ── Tokenization ────────────────────────────────────────────────── */
function tokenize(text: string): string[] {
  const lower = text.toLowerCase()
    .normalize('NFD').replace(/[̀-ͯ]/g, '')   // strip accents
    .replace(/[^a-z0-9\s]/g, ' ')
  const words = lower.split(/\s+/).filter(w => w.length > 2 && !SPANISH_STOP_WORDS.has(w))
  // unigrams + bigrams
  const tokens = [...words]
  for (let i = 0; i < words.length - 1; i++) {
    tokens.push(`${words[i]}_${words[i + 1]}`)
  }
  return tokens
}

/* ── TF-IDF ──────────────────────────────────────────────────────── */
interface Vector { [term: string]: number }

function tfidfVectors(documents: string[]): Vector[] {
  const docTokens = documents.map(tokenize)
  const n = documents.length

  // Document frequency
  const df: Record<string, number> = {}
  for (const tokens of docTokens) {
    const seen = new Set<string>()
    for (const t of tokens) {
      if (seen.has(t)) continue
      seen.add(t)
      df[t] = (df[t] ?? 0) + 1
    }
  }

  // IDF
  const idf: Record<string, number> = {}
  for (const t of Object.keys(df)) {
    idf[t] = Math.log((n + 1) / (df[t] + 1)) + 1   // smoothed
  }

  // TF-IDF vectors with L2 normalization
  return docTokens.map(tokens => {
    const tf: Record<string, number> = {}
    for (const t of tokens) tf[t] = (tf[t] ?? 0) + 1
    // sublinear tf
    for (const t of Object.keys(tf)) tf[t] = 1 + Math.log(tf[t])

    const vec: Vector = {}
    let norm = 0
    for (const t of Object.keys(tf)) {
      const w = tf[t] * (idf[t] ?? 0)
      vec[t] = w
      norm += w * w
    }
    norm = Math.sqrt(norm) || 1
    for (const t of Object.keys(vec)) vec[t] /= norm
    return vec
  })
}

function cosine(a: Vector, b: Vector): number {
  // Iterate over the smaller vector for efficiency
  const [small, big] = Object.keys(a).length < Object.keys(b).length ? [a, b] : [b, a]
  let dot = 0
  for (const t of Object.keys(small)) {
    if (big[t] !== undefined) dot += small[t] * big[t]
  }
  return dot   // vectors are L2-normalized → dot product is cosine
}

/* ── Main clustering function ────────────────────────────────────── */
export function clusterArticles(
  articles: ArticleInput[],
  threshold = 0.18,
): ClusterResult[] {
  if (articles.length < 2) return []

  const texts = articles.map(a => `${a.title} ${a.summary}`)
  const vectors = tfidfVectors(texts)
  const n = articles.length
  const visited = new Array<boolean>(n).fill(false)
  const clusters: ClusterResult[] = []

  for (let i = 0; i < n; i++) {
    if (visited[i]) continue
    const indices = [i]
    visited[i] = true

    for (let j = i + 1; j < n; j++) {
      if (visited[j]) continue
      const sim = cosine(vectors[i], vectors[j])
      if (sim >= threshold) {
        // Only cluster if source isn't already represented
        const sourcesIn = new Set(indices.map(k => articles[k].sourceSlug))
        if (!sourcesIn.has(articles[j].sourceSlug)) {
          indices.push(j)
          visited[j] = true
        }
      }
    }

    const sourceSlugs = indices.map(k => articles[k].sourceSlug)
    if (new Set(sourceSlugs).size >= 2) {
      const scores: Record<number, number> = {}
      for (const k of indices) scores[articles[k].id] = cosine(vectors[i], vectors[k])

      // Representative title: longest
      const repIdx = indices.reduce((a, b) =>
        articles[a].title.split(' ').length >= articles[b].title.split(' ').length ? a : b
      )
      clusters.push({
        articleIds: indices.map(k => articles[k].id),
        sourceSlugs,
        representativeTitle: articles[repIdx].title,
        similarityScores: scores,
      })
    }
  }

  console.log(`Clustered ${articles.length} articles into ${clusters.length} clusters`)
  return clusters
}
